shell.nix

```diff
@@ -0,0 +1,39 @@
+{ nixpkgs ? import <nixpkgs> {} }:
+
+let
+  haskellPackages = nixpkgs.haskellPackages;
+
+  hadrianPackages = nixpkgs.haskellPackages.override {
+    overrides = self: super: let
+      in {
+        hadrian = self.callCabal2nix "hadrian" ./. {};
+        shake = self.callHackage "shake" "0.16" {};
+        Cabal = self.callCabal2nix "Cabal" ./../libraries/Cabal/Cabal {};
+        filepath = self.callCabal2nix "filepath" ./../libraries/filepath {};
+        text = self.callCabal2nix "text" ./../libraries/text {} ;
+        hpc = self.callCabal2nix "hpc" ./../libraries/hpc {};
+        parsec = self.callCabal2nix "parsec" ./../libraries/parsec {};
+        HUnit = nixpkgs.haskell.lib.dontCheck (self.callHackage "HUnit" "1.3.1.2" {});
+      }; };
+
+  hadrian = hadrianPackages.ghcWithPackages (ps: [ ps.hadrian ]);
```

Comment:
I'm not familiar with nix, so perhaps this is the second part of the puzzle -- is this the nix-based build script for Hadrian?

Reply:
It overrides the default haskell packageset by replacing hadrian, shake, etc by the local ones, except for shake for which we specifically require 0.16, and HUnit 1.3.1.2. Finally, line 19 uses this "new" package set to define a build environment that has a ghc + hadrian and all of its dependencies in it. This is then used along with many other things to provision the entire build environment needed to build _ghc_.

Reply:
So, if the user edits `hadrian/UserSettings.hs` will the invokation of the `build-nix` script automatically rebuild Hadrian?

Reply:
Yes. In fact, it will even do it when any file from the hadrian dir changes, but this is a quirk @mpickering told me about and we know how to fix it. If @mpickering doesn't get around to doing it in the upcoming days I'll do it.

Reply:
@alpmestan Note that in principle Hadrian should also be rebuilt when the in-tree Cabal library is modified. But this is a rather rare scenario.

Reply:
That is also taken care of correctly but not evident from the script. 

Essentially `callCabal2nix` creates a function which takes all the dependencies of `hadrian` and returns a recipe to construct `hadrian` using these dependencies. It then performs some magic to automatically fill in these arguments correctly and because we specify that we want to use the in-tree `Cabal` on a later line, it passes this version of `Cabal` as an argument. The key principle of `nix` is that when an argument changes, then you have to rebuild the result, so when the in-tree `Cabal` changes, then `Cabal` and hence `hadrian` will be rebuilt.

Reply:
Aga, that's cool!

Reply:
We got you halfway to using nix haven't we? :)

Reply:
@alpmestan Yes, you have :) I've already heard a lot of good things about nix and I'm tempted!